src/array.c

```diff
@@ -611,10 +636,22 @@ JL_DLLEXPORT void jl_arrayset(jl_array_t *a JL_ROOTING_ARGUMENT, jl_value_t *rhs
         else {
             hasptr = a->flags.hasptr;
         }
-        arrayassign_safe(hasptr, jl_array_owner(a), &((char*)a->data)[i * a->elsize], rhs, a->elsize);
+        if (order == jl_memory_order_notatomic)
+            arrayassign_safe(hasptr, jl_array_owner(a), &((char*)a->data)[i * a->elsize], rhs, a->elsize);
+        else
+        {
+            if ((a->elsize & (a->elsize - 1)) != 0 || a->elsize > MAX_POINTERATOMIC_SIZE)
+                jl_error("atomic_arrayset: invalid pointer for atomic operation");
```

Comment:
Can you add a section to https://hackmd.io/s/SyFljvtdO describing all of the intended new APIs here and how we are intending to deal with the size and alignment problems? I see you've done some effort here to account for that, similar to the raw pointer API, but I think it would be good to be clear about the tradeoffs we made in our documentation.